src/nupic/utils/SlidingWindow.hpp

```diff
@@ -0,0 +1,191 @@
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses.
+ *
+ * http://numenta.org/licenses/
+ * ---------------------------------------------------------------------
+ */
+
+
+#ifndef NUPIC_UTIL_SLIDING_WINDOW_HPP
+#define NUPIC_UTIL_SLIDING_WINDOW_HPP
+
+#include <vector>
+#include <algorithm>
+#include <iterator>
+#include <cmath>
+#include <string>
+
+#include <nupic/types/Types.hpp>
+#include <nupic/utils/Log.hpp>
+#include <nupic/math/Math.hpp> //for macro ASSERT_INPUT_ITERATOR
+#include <nupic/utils/VectorHelpers.hpp> //printVector
+
+using nupic::utils::VectorHelpers;
+
+namespace nupic {
+  namespace util {
+
+template<class T> 
+class SlidingWindow {
+  public:
+    SlidingWindow(UInt maxCapacity, std::string id="SlidingWindow", int debug=0) : 
+      maxCapacity(maxCapacity),
+      ID(id),
+      DEBUG(debug)
+{
+      buffer_.reserve(maxCapacity);
+      idxNext_ = 0;
+} 
+
+
+    template<class IteratorT> 
+    SlidingWindow(UInt maxCapacity, IteratorT initialData_begin, 
+      IteratorT initialData_end, std::string id="SlidingWindow", int debug=0): 
+      SlidingWindow(maxCapacity, id, debug) {
+      // Assert that It obeys the STL forward iterator concept
+      ASSERT_INPUT_ITERATOR(IteratorT);
+      for(IteratorT it = initialData_begin; it != initialData_end; ++it) {
+        append(*it);
+      }
+      idxNext_ = buffer_.size() % maxCapacity; //TODO remove, not needed
+    }
+
+    const UInt maxCapacity;
+    const std::string ID; //name of this object
+    const int DEBUG;
+
+    size_t size() const {
+      NTA_ASSERT(buffer_.size() <= maxCapacity);
+      return buffer_.size();
+    }
+
+
+    /** append new value to the end of the buffer and handle the 
+       "overflows"-may pop the oldest element if full. 
+      */
+    void append(T newValue) {
+      if(size() < maxCapacity) {
+        buffer_.push_back(newValue);
+      } else {
+        buffer_[idxNext_] = newValue;
+      }
+      //the assignment must be out of the [] above, so not [idxNext_++%maxCap],
+      // because we want to store the value %maxCap, not only ++
+      idxNext_ = (idxNext_ +1 ) %maxCapacity;
+    }
+
+
+      /** like append, but return the dropped value if it was dropped.
+        :param T newValue - new value to append to the sliding window
+        :param T& - a return pass-by-value with the removed element,
+          if this function returns false, this value will remain unchanged.
+        :return bool if some value has been dropped (and updated as 
+          droppedValue) 
+      */
+      bool append(T newValue, T& droppedValue) {
```

Comment:
I mentioned this above, but I still think this should be a `T*`. It's fundamentally the same, but it makes the code much more readable.